### DISCUSSION TITLE
Providing redirect and rewrite for trailing slash URL's

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#30660] [Ideas]
### Describe the feature you'd like to request

Currently, in NextJS the trailing slash redirect seems to be taking priority over the custom redirects mentioned in next.config.js. I would like to prioritize user mentioned redirect and rewrites over the nextjs defaults.

### Describe the solution you'd like

The solution I like is simple, suppose I mention a redirect or rewrite of "/abc/" path to "/xyz" then it should take priority over the NextJS auto-redirect of trailing slash, I ended up looking in NextJS framework files and found that in "lib/load-custom-routes.js" file changing the "redirect.unshift" to "redirect.push" can do the trick but I don't know if it creates a breaking change in other parts as well. 

### Describe alternatives you've considered

Alternatively, what I considered the solution is to let the NextJS auto-redirect do its thing and mention the redirect or rewrite for the final path. For example NextJS redirect "/abc/" to "/abc", then I mention redirect or rewrite for "/abc" to "/xyz". This seems to be a good solution but it is creating redirect hops which isn't very considerable. Also for some edge cases, we may need to render some other content on /abc/ route which becomes impossible unless` trailingSlash:true` is mentioned in next.config.js but then we end up all the routes ending with "/".

**Comment by donnersvensson** [comment]
Is there anything planned to implement this feature? We would welcome it a lot. When relaunching a big website we have to do a lot of redirects and we end up with a lot redirect chains ("/abc/" to "/abc" to "/xyz"). It´s harder to maintain such a redirect file and it is not as good for SEO as a simple redirect would be.

**Comment by kevinmoch-qz** [comment]
Commenting to bring more attention to this question. We'relso running into a situation where being unable to prioritize our own re-writes before the trailing slash redirect is leading to severe disadvantages for SEO.

I've also been through [this issue](https://github.com/vercel/next.js/discussions/23988) on expanding the options for the trailing slash config in `next.config.js`, but that's been open for years with no word that any updates are planned. Hoping for either this or that to be prioritized sometime soon.

**Reply by kevinmoch-qz** [reply]
FWIW, the way we ended up etting around this for the present was to remove the trailing slash in our custom server before passing the routing off to Next.

**Comment by samparmenter** [comment]
This is something we are running into as well. Exactly as the above. Our redirects are leading to a chain of trailing slash -> non-trailing slash -> final url. 

`/a/` -> `/a` -> `/b` instead of just `/a/` -> `/b`
